src/keys.rs

```diff
@@ -27,19 +38,58 @@ impl KeyPair {
     }
 
     pub fn public(&self) -> PublicKey {
-        self.public_key
+        self.public_key.clone()
+    }
+
+    pub fn sign(&self, digest: SigHash) -> Signature {
+        // TODO: Use a sensible tag
+        let ecdsa = ECDSA::from_tag(b"my-tag").enforce_low_s();
+
+        ecdsa.sign(
+            &self.secret_key.0,
+            &digest.into_inner(),
+            Derivation::Deterministic,
+        )
+    }
+
+    pub fn presign(&self, Y: PublishingPublicKey, digest: SigHash) -> EncryptedSignature {
+        // TODO: Use a sensible tag
+        let adaptor = Adaptor::from_tag(b"my-tag");
+
+        adaptor.encrypted_sign(
+            &self.secret_key.0,
+            &(Y.0).0,
+            &digest.into_inner(),
+            Derivation::Deterministic,
+        )
+    }
+}
+
+impl SecretKey {
+    pub fn public(&self) -> PublicKey {
+        let sk = &self.0;
+        // TODO: Determine what is the correct marker for this public key
+        let public_key = g!(sk * G).mark::<Normal>();
+
+        PublicKey(public_key)
     }
 }
 
 impl From<SecretKey> for KeyPair {
     fn from(secret_key: SecretKey) -> Self {
         Self {
-            secret_key,
-            public_key: PublicKey::from_secret_key(&SECP, &secret_key),
+            secret_key: secret_key.clone(),
+            public_key: secret_key.public(),
         }
     }
 }
 
+impl From<PublicKey> for Point {
```

Comment:
I think it would be nice if it could just be `PublishingPublicKey(Point)` rather than having an intermediate newtype.
And to have just `Scalar` and `Point` in KeyPair. 

Are `SecretKey` and `PublicKey` necessary? Is there anything I can provide to help remove them?

Reply:
I agree :+1: What do you think about introducing a wrapper type for `x` and `X`, such as `Ownership{Secret, Public}Key`? Would that name make sense?

Reply:
> Is there anything I can provide to help remove them?

There's one key aspect: conversion to and from `bitcoin::secp256k1::PublicKey`.

Reply:
For now I've implemented `TryFrom` on our types.

Reply:
Seeing as there are a lot of keys flying around it might make sense to have them. I usually introduce new types as late as possible otherwise I spend too much time maintaining them.